src/main/flight/pid.c

```diff
@@ -538,6 +567,79 @@ static void handleItermRotation()
     }
 }
 
+#ifdef USE_ACRO_TRAINER
+
+int8_t acroTrainerSign(float x)
+{
+    return (x > 0.0f) - (x < 0.0f);
+}
+
+// Acro Trainer - Manipulate the setPoint to limit axis angle while in acro mode
+// There are three states:
+// 1. Current angle has exceeded limit
+//    Apply correction to return to limit (similar to pidLevel)
+// 2. Future overflow has been projected based on current angle and gyro rate
+//    Manage the setPoint to control the gyro rate as the actual angle  approaches the limit (try to prevent overshoot)
+// 3. If no potential overflow is detected, then return the original setPoint
+
+static float applyAcroTrainer(int axis, const rollAndPitchTrims_t *angleTrim, float setPoint)
+{
+    float ret = setPoint;
+
+#ifndef SIMULATOR_BUILD
+    if (!FLIGHT_MODE(ANGLE_MODE) && !FLIGHT_MODE(HORIZON_MODE)) {
+        bool resetIterm = false;
+        float projectedAngle;
+        const int8_t setpointSign = acroTrainerSign(setPoint);
+        const float currentAngle = (attitude.raw[axis] - angleTrim->raw[axis]) / 10.0f;
+        const int8_t angleSign = acroTrainerSign(currentAngle);
+
+        // Limit and correct the angle when it exceeds the limit
+        if ((fabsf(currentAngle) > acro_trainer_angle_limit) || (acroTrainerAxisState[axis] != 0)) {
+            if (angleSign == -setpointSign) {  // stick has reversed - stop limiting
+                acroTrainerAxisState[axis] = 0;
+            } else {
+                if (acroTrainerAxisState[axis] != angleSign) {
+                    acroTrainerAxisState[axis] = angleSign;
+                    resetIterm = true;
+                }
+                ret = ((acro_trainer_angle_limit * angleSign) - currentAngle) * ACRO_TRAINER_GAIN;
+            }
+        }
+        
+        // Not currently over the limit so project the angle based on current angle and
+        // gyro angular rate using a sliding window based on gyro rate (faster rotation means larger window.
+        // If the projected angle exceeds the limit then apply limiting to minimize overshoot.
+        if (acroTrainerAxisState[axis] == 0) {
```

Comment:
There is discontinuity when crossing limit, and controller output decreases - actual angle is used instead of projected
This is probably not desirable, IMO pre-cross controller should blend `setPoint` so that it is 0 at crossover.
Angle rate PID should handle gyro rate ...

Reply:
Well the angle holding logic is borrowed directly from `pidLevel()` and uses the same approach.  An angle error times gain to determine the new setpoint.  The difference is that the angle is compared to the limit angle rather than zero (for level modes).

I tried using the projected angle after crossover in some of my many different versions.  I found it didn't work well and added more oscillation because of the phase shift introduced by the look-ahead period.  I even had a version that blended a look-ahead window initially and diminished it to zero and even though you'd think that would work in practice it didn't.  Still might be some opportunities to improve but I've been spinning my wheels trying to make it "perfect" and over-engineer when at this point it just needs to be functional.  Had to go back to the philosophy that flying on the angle limit is not the goal - rather it's to help pilots learn to control the craft so they don't have to rely on the limit.

Reply:
See above ... 

Did you try simply calculating setpoint from over-limit case and limiting user setpoint to it? With large user setpoint, it should behave exactly as limit mode, with smaller user input it will have no influence.
Tuned angle-rate PID should take care of actual gyro rate

Reply:
BTW: using pidLevel user coefficients may be an option ...